File: internal/boveda/identidad.go
package boveda

// La identidad de una bóveda para compartir copias (ADR 0043).
//
// **Una semilla de 32 bytes dentro del cuerpo cifrado, y todo lo demás se deriva
// de ella.** No se guarda ninguna llave: derivarlas cuesta microsegundos y una
// llave guardada es una llave que se puede desincronizar de su semilla.
//
// De la semilla salen dos cosas, con etiquetas distintas para que no puedan
// confundirse nunca:
//
//   - **X25519**, para que otros cifren hacia ti (HPKE, RFC 9180).
//   - **Ed25519**, para firmar lo que mandas.
//
// **Se crea una sola vez y no cambia**: cambiarla es cambiar de identidad, y lo
// que te mandaron antes se queda sin abrir.

import (
	"crypto/ecdh"
	"crypto/ed25519"
	"crypto/hkdf"
	"crypto/sha256"
	"encoding/base64"
	"encoding/json"
	"errors"
	"time"

	"github.com/webcafeina/esfinge/internal/cripto"
)

// La semilla se guarda en base64url sin relleno, como la clave de bóveda.
var b64 = base64.RawURLEncoding

// Suite es el conjunto de HPKE con el que se cifra hacia esta identidad.
//
// Viaja como campo —y no como una constante del código— para que el día que haya
// ML-KEM razonable en el navegador se pueda añadir sin romper lo ya mandado
// (ADR 0043). Hoy solo hay una, y es la que el navegador puede hacer con una
// biblioteca pequeña.
const Suite = "DHKEM(X25519)/HKDF-SHA256/ChaCha20-Poly1305"

const (
	infoCifrado = "esfinge/identidad/cifrado/v1"
	infoFirma   = "esfinge/identidad/firma/v1"
)

// identidad es lo que se guarda: la semilla y cuándo nació.
//
// `Creada` no es adorno: es lo que desempata si dos equipos crearan una identidad
// cada uno antes de verse (ADR 0043).
type identidad struct {
	Semilla string `json:"semilla"` // 32 bytes en base64url
	Creada  string `json:"creada"`
	Suite   string `json:"suite"`
}

// Identidad es lo que se puede enseñar de una identidad: sus llaves públicas y su
// huella. **La semilla no sale de aquí.**
type Identidad struct {
	Suite   string `json:"suite"`
	Cifrado []byte `json:"cifrado"` // X25519, 32 bytes
	Firma   []byte `json:"firma"`   // Ed25519, 32 bytes
	Huella  string `json:"huella"`
}

var errSinIdentidad = errors.New("Esta bóveda todavía no tiene identidad")

// derivar saca de la semilla las dos llaves privadas.
func derivarIdentidad(semilla []byte) (*ecdh.PrivateKey, ed25519.PrivateKey, error) {
	deCifrado, err := hkdf.Key(sha256.New, semilla, nil, infoCifrado, 32)
	if err != nil {
		return nil, nil, err
	}
	deFirma, err := hkdf.Key(sha256.New, semilla, nil, infoFirma, 32)
	if err != nil {
		return nil, nil, err
	}
	cifrado, err := ecdh.X25519().NewPrivateKey(deCifrado)
	if err != nil {
		return nil, nil, err
	}
	return cifrado, ed25519.NewKeyFromSeed(deFirma), nil
}

// HuellaDeIdentidad es lo que se compara por teléfono: siete grupos de cuatro
// símbolos del alfabeto de la clave de recuperación, que está pensado para leerse
// en voz alta sin confundir la I con el 1 ni la O con el 0.
//
// Cubre **la suite y las dos llaves**: cambiar cualquiera de las tres cambia la
// huella, que es justo lo que hace falta para que comparar sirva de algo.
func HuellaDeIdentidad(suite string, cifrado, firma []byte) string {
	h := sha256.New()
	h.Write([]byte(suite))
	h.Write([]byte{0})
	h.Write(cifrado)
	h.Write([]byte{0})
	h.Write(firma)
	return agrupar(aPalabras(h.Sum(nil))[:28], 4)
}

// aPalabras es la misma conversión que la clave de recuperación: cinco bits por
// símbolo, sin I, L, O ni U.
func aPalabras(b []byte) string {
	var sb []byte
	var acumulado, bits uint32
	for _, x := range b {
		acumulado = acumulado<<8 | uint32(x)
		bits += 8
		for bits >= 5 {
			bits -= 5
			sb = append(sb, alfabeto[(acumulado>>bits)&31])
		}
	}
	return string(sb)
}

func agrupar(s string, cada int) string {
	var sb []byte
	for i, r := range s {
		if i > 0 && i%cada == 0 {
			sb = append(sb, '-')
		}
		sb = append(sb, byte(r))
	}
	return string(sb)
}

// Identidad devuelve la identidad pública de esta bóveda, creándola si todavía no
// la tiene. **Guarda si la crea**, porque una identidad que no llega al disco es
// una identidad distinta en el próximo arranque.
func (b *Boveda) Identidad() (Identidad, error) {
	b.mu.Lock()
	defer b.mu.Unlock()
	if b.llave == nil {
		return Identidad{}, ErrCerrada
	}
	if b.cont.Identidad == nil {
		semilla, err := cripto.Azar(32)
		if err != nil {
			return Identidad{}, err
		}
		b.cont.Identidad = &identidad{
			Semilla: b64.EncodeToString(semilla),
			Creada:  ahora().UTC().Format(time.RFC3339),
			Suite:   Suite,
		}
		b.cuerpoSucio = true
		if err := b.guardar(); err != nil {
			b.cont.Identidad = nil
			return Identidad{}, err
		}
	}
	return publicaDe(b.cont.Identidad)
}

func publicaDe(i *identidad) (Identidad, error) {
	if i == nil {
		return Identidad{}, errSinIdentidad
	}
	semilla, err := b64.DecodeString(i.Semilla)
	if err != nil || len(semilla) != 32 {
		return Identidad{}, errors.New("La identidad de la bóveda no se entiende")
	}
	cifrado, firma, err := derivarIdentidad(semilla)
	if err != nil {
		return Identidad{}, err
	}
	pubCifrado := cifrado.PublicKey().Bytes()
	pubFirma := []byte(firma.Public().(ed25519.PublicKey))
	suite := i.Suite
	if suite == "" {
		suite = Suite
	}
	return Identidad{
		Suite:   suite,
		Cifrado: pubCifrado,
		Firma:   pubFirma,
		Huella:  HuellaDeIdentidad(suite, pubCifrado, pubFirma),
	}, nil
}

// fundirIdentidad elige una de las dos, y **siempre la misma la mire quien la
// mire**: la más antigua, y si empatan la de semilla menor. Solo puede haber dos
// si dos equipos crearon la suya antes de verse.
func fundirIdentidad(l, r *identidad) *identidad {
	switch {
	case l == nil:
		return r
	case r == nil:
		return l
	case l.Creada != r.Creada:
		if l.Creada < r.Creada {
			return l
		}
		return r
	case l.Semilla <= r.Semilla:
		return l
	default:
		return r
	}
}

// comoJSON es lo que se escribe en el cuerpo. Se usa desde las pruebas cruzadas.
func (i *identidad) comoJSON() json.RawMessage {
	b, _ := json.Marshal(i)
	return b
}

File: navegador/src/nucleo/identidad.ts
/**
 * La identidad de la bóveda para compartir copias, en TypeScript (ADR 0043).
 *
 * **El espejo de `internal/boveda/identidad.go`**, y como todo lo de `nucleo/`,
 * tiene que dar exactamente lo mismo: las mismas llaves y la misma huella para la
 * misma semilla. Lo vigilan las pruebas cruzadas.
 *
 * Se usa WebCrypto y **ninguna biblioteca nueva**: X25519 y Ed25519 están en
 * `crypto.subtle` de los navegadores a los que va la extensión —comprobado en el
 * Chromium de las pruebas, y Firefox se pide 140 o más—. La única maña que hace
 * falta es que WebCrypto no deriva la llave pública de una privada: se importa la
 * privada y se exporta como JWK, que trae la pública en `x`.
 */

const utf8 = new TextEncoder();

/** El mismo conjunto que Go: viaja como campo para poder cambiarlo algún día. */
export const SUITE = "DHKEM(X25519)/HKDF-SHA256/ChaCha20-Poly1305";

const INFO_CIFRADO = "esfinge/identidad/cifrado/v1";
const INFO_FIRMA = "esfinge/identidad/firma/v1";

/** El alfabeto de la clave de recuperación: 32 símbolos sin I, L, O ni U. */
const ALFABETO = "0123456789ABCDEFGHJKMNPQRSTVWXYZ";

export type Identidad = {
  suite: string;
  /** X25519, 32 bytes. */
  cifrado: Uint8Array;
  /** Ed25519, 32 bytes. */
  firma: Uint8Array;
  huella: string;
};

/** Lo que se guarda en el cuerpo de la bóveda, en la sección `identidad`. */
export type IdentidadGuardada = { semilla: string; creada: string; suite?: string };

async function hkdf(semilla: Uint8Array, info: string): Promise<Uint8Array> {
  const k = await crypto.subtle.importKey("raw", new Uint8Array(semilla), "HKDF", false, ["deriveBits"]);
  const bits = await crypto.subtle.deriveBits(
    { name: "HKDF", hash: "SHA-256", salt: new Uint8Array(0), info: utf8.encode(info) },
    k,
    256,
  );
  return new Uint8Array(bits);
}

/**
 * La llave pública de una privada de 32 bytes.
 *
 * WebCrypto no sabe «dame la pública de esta privada», así que se envuelve la
 * privada en el PKCS#8 mínimo —que para estas dos curvas es una cabecera fija más
 * los 32 bytes— y se exporta como JWK, que trae la pública en `x`.
 */
async function publicaDe(privada: Uint8Array, alg: "X25519" | "Ed25519"): Promise<Uint8Array> {
  const oid = alg === "X25519" ? 0x6e : 0x70;
  const pkcs8 = new Uint8Array([
    0x30, 0x2e, 0x02, 0x01, 0x00, 0x30, 0x05, 0x06, 0x03, 0x2b, 0x65, oid, 0x04, 0x22, 0x04, 0x20,
    ...privada,
  ]);
  const usos: KeyUsage[] = alg === "X25519" ? ["deriveBits"] : ["sign"];
  const k = await crypto.subtle.importKey("pkcs8", pkcs8, { name: alg }, true, usos);
  const jwk = (await crypto.subtle.exportKey("jwk", k)) as { x?: string };
  if (!jwk.x) throw new Error(`WebCrypto no ha dado la llave pública de ${alg}`);
  return desdeBase64url(jwk.x);
}

function desdeBase64url(s: string): Uint8Array {
  const b = atob(s.replace(/-/g, "+").replace(/_/g, "/"));
  return Uint8Array.from(b, (c) => c.charCodeAt(0));
}

/** Cinco bits por símbolo, como la clave de recuperación y como `aPalabras` en Go. */
function aPalabras(b: Uint8Array): string {
  let salida = "";
  let acumulado = 0;
  let bits = 0;
  for (const x of b) {
    acumulado = ((acumulado << 8) | x) >>> 0;
    bits += 8;
    while (bits >= 5) {
      bits -= 5;
      salida += ALFABETO[(acumulado >>> bits) & 31];
    }
  }
  return salida;
}

/**
 * La huella que se compara por teléfono: siete grupos de cuatro símbolos que
 * cubren la suite y las dos llaves. Igual que `HuellaDeIdentidad` en Go.
 */
export async function huellaDeIdentidad(suite: string, cifrado: Uint8Array, firma: Uint8Array): Promise<string> {
  const partes = [utf8.encode(suite), new Uint8Array([0]), cifrado, new Uint8Array([0]), firma];
  const total = partes.reduce((n, p) => n + p.length, 0);
  const todo = new Uint8Array(total);
  let i = 0;
  for (const p of partes) {
    todo.set(p, i);
    i += p.length;
  }
  const h = new Uint8Array(await crypto.subtle.digest("SHA-256", todo));
  const palabras = aPalabras(h).slice(0, 28);
  return (palabras.match(/.{1,4}/g) ?? []).join("-");
}

/** La identidad pública que sale de una semilla de 32 bytes. */
export async function identidadDeSemilla(semilla: Uint8Array, suite = SUITE): Promise<Identidad> {
  if (semilla.length !== 32) throw new Error("La semilla de la identidad tiene que ser de 32 bytes");
  const cifrado = await publicaDe(await hkdf(semilla, INFO_CIFRADO), "X25519");
  const firma = await publicaDe(await hkdf(semilla, INFO_FIRMA), "Ed25519");
  return { suite, cifrado, firma, huella: await huellaDeIdentidad(suite, cifrado, firma) };
}

/**
 * De las dos, la misma que elegiría Go: la más antigua, y si empatan la de semilla
 * menor. Solo hay dos si dos equipos crearon la suya antes de verse.
 */
export function fundirIdentidad(
  l: IdentidadGuardada | undefined,
  r: IdentidadGuardada | undefined,
): IdentidadGuardada | undefined {
  if (!l) return r;
  if (!r) return l;
  if (l.creada !== r.creada) return l.creada < r.creada ? l : r;
  return l.semilla <= r.semilla ? l : r;
}
